### ISSUE TITLE
[BUG] Too many caches in GitHub Actions CI

### ISSUE THREAD

**Issue by adamretter** [eXist-db/exist#4861]
At present we have a number of different caches used across several GitHub Actions workflows. In terms of cache size, the most significant one is that for caching Maven dependencies (`~/.m2/repository`).

The caching of Maven repositories is performed for by these workflows:

1. `ci-deploy.yml`
        * Target: The `master` and `develop` branches, and any PR's opened against them.
	* Jobs:
		1. Build and Test Images
	* Cache key: `${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}`
	* Restore key: `${{ runner.os }}-m2`

2. `ci-test.yml`
         * Target: All branches, and all PRs (ignoring the Code Coverage step which only executes for the `develop` branch).
	* Jobs:
		1. License check
		2. Dependency checks
		5. Javadoc
		6. Test
	* Cache key: `setup-java-${{ platform }}-maven-${{ hashFiles('**/pom.xml') }}`
	* Restore key: `setup-java-${{ platform }}-maven-${{ hashFiles('**/pom.xml') }}`

3. `ci-xqts.yml`
         * Target: All branches, and all PRs.
	* Jobs:
		1. W3C XQuery Test Suite
	* Cache key: `setup-java-${{ platform }}-maven-${{ hashFiles('**/pom.xml') }}`
	* Restore key: `setup-java-${{ platform }}-maven-${{ hashFiles('**/pom.xml') }}`

4. `sonarcloud.yml`
         * Target: The `master` and `develop` branches, and all PRs.
	* Jobs:
		1. SonarCloud Analysis
	* Cache key: `${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}`
	* Restore key: `${{ runner.os }}-m2`

## Problems with this approach

1. The `ci-deploy.yml`, and the `sonarcloud.yml` workflows use the same cache key, which means that a PR that changes anything in any `pom.xml` will cause a replacement of that cache by both workflows. This in turn leads to a race-condition - whichever workflow runs last will win (i.e. have its cache persisted). The problem is that each workflow will produce a different set of Maven dependencies (i.e. different set of cacheable dependencies), as they execute different parts of the build. Therefore each time the cache is retrieved, if a `pom.xml` file was changed, then the cache *may* be invalid to both workflows and thus pointless; it may actually slow down the build!

2. The `ci-test.yml` and the `ci-xqts.yml` also share a cache key, but will generate different caches. They have the same problem discussed above in (1).

3. We have many PRs that come through that generate a change to a single dependency in a single `pom.xml`. These are autogenerated by @dependabot-bot. Using @dependabot-bot is great, but due to how we have configured our Caches for Maven (i.e. the `actions/setup-java@v3` GitHub Action with `cache: 'maven` in `ci-test.yml`), each @dependabot-bot will create a new Cache.

4. Each of our caches for Maven Dependencies is ~360MB at present, GitHub limits the amount of cache space for a project to a generous 10GB, but we frequently hit this ceiling.  <img width="897" alt="Screenshot 2023-04-10 at 10 16 40" src="https://user-images.githubusercontent.com/1264057/230872960-9b0a96cd-cfbd-47db-a856-35b6246165f9.png">

## Insights

1. PRs generated for @dependabot-bot, will each create a Cache that can only be used by that PR, and then if merged: subsequent builds of the `develop` branch, and any later PRs to `develop` (that don't change any `pom.xml)`. Each of these Caches is expensive in terms of disk-space (~360MB at present).

2. The space available for Caches is limited to 10GB, each cache joins a queue for eviction when the 10GB disk-space is exceeded. This queue is FIFO based and not an LRU. This means that if a great deal of the total 10GB of available cache space is taken up with @dependabot-bot caches (that may only be used twice), and is interspersed with caches that are used more frequently, then these @dependabot-bot caches may cause more frequently used caches to be invalidated sooner. A future workflow that would have used that "hotter" cache, will have to re-download these dependencies and create a new Cache; thus slowing down that workflow.

3. The cache keys used for the Maven Caches include `${{ platform }}` component. However as this is Java, there are very little in the way of platform specific dependencies, and so this could be removed which would reduce the number of caches by 3x (our platforms: Linux, macOS, Windows).

4. The majority of changes to the `pom.xml` files come from Maven dependency updates courtesy of @dependabot-bot PRs.

5. The majority of non-dependabot PRs do not change the `pom.xml` files.

6. The majority of PRs against a branch, make either zero or one change (e.g. @dependabot-bot) to the dependencies in a `pom.xml` file. If that PR could reuse a cache generated for the base-branch then that cache would already contain > 99% of the required dependencies.

7.  The cache for a branch need only be replaced after that branch has a change (PR) merged that adds/updates/changes a dependency (i.e. pom.xml) file.

8. Instead of using the `cache: maven` property of the GitHub Action `actions/setup-java@v3` which will cache and match on exact cache keys, we can instead directly use the `key` and `restore-keys` properties of the `actions/cache@v3` GitHub Action to potentially read and write from different caches.

9. The `actions/cache@v3` GitHub Action provides a granularity of configuration that allows us to have some workflow jobs that only read from the cache, whilst others may write to the cache, or both.

## Proposed Solution

By considering:
1. The lifetime of a Cache of Maven Dependencies
2. The clients (workflows) that need to work with them
3. The isolation of dependencies between those clients
4. If each client needs read and/or write access to the cache

I propose the following policy for our GitHub CI Maven caches:
1. One Maven Cache per Git branch.
Due to our development model, of "fork and send a PR", we really only have 5 active base-branches: `master, develop, develop-6.x.x, develop-5.x.x, develop-4.x.x`. It is likely that the number of active branches will decrease in the near future as we stop providing bugfixes to older version of our product.
2. PRs do not create Caches.
A PR is taken from a base-branch, when the GitHub Actions workflows execute for a PR, it can read the Cache for a base-branch. That cache will either be 100% complete already, or if the PR changes a few dependencies, it will be very close to 100% complete. A GitHub workflow for a PR never writes a Cache for Maven dependencies.
3. Maven Caches created by Merge of PR.
A GitHub Actions Workflow that runs after a PR has been merged to a base-branch may create a new Maven Cache. The Cache key must be named based on the base-branch name, so it may be used by subsequent workflows on that branch and new PRs against that branch. For example:
    * Cache key: `maven-cache-${{github.base_ref}}-${{ hashFiles('**/pom.xml') }}`
    * Restore keys:
        1. `maven-cache-${{github.base_ref}}-${{ hashFiles('**/pom.xml') }}`
        2. `maven-cache-${{github.base_ref}}-`
        3. `maven-cache-`

NOTE: It is tempting to think about removing the `${{github.base_ref}}` from the Cache Key. The issue with doing that is that each of our active branches represent major differences in the product (i.e. huge dependency changes), if we were to cache and restore against just the key `maven-cache-${{github.base_ref}}`, we would have very little chance of taking the cache that is most correct for the major product version.

**Comment by duncdrum** [comment]
see [Community Call Monday, January 26, 2026](https://docs.google.com/document/d/15b2cnXRmD06-jQepLas6fUbiHs4U-SFqCqfu_sCDGIk/edit?usp=sharing)

**Comment by dizzzz** [comment]
See also notes of today :-)  https://docs.google.com/document/d/1rf_Q8lQ2zSXLU4-swoxBq5UNGGASpFGvPP10FKGkT3o/edit?tab=t.0

**Comment by dizzzz** [comment]
So the idea is to reduce the nr of caches. In the teleconference we had a brainstorm session.  Ultimate goal is to reduce the number of downloads as much as possible, as where 'deltas' in dependencies in PRs are not relevant. Much is inspired by previous text of Adam.

So the idea is to have only a limited set of caches. Caches will be only filled/written by the main branches: master, develop, develop-6.x.x (and other develop-xxxx branches when needed).

For PRs/branches a copy shall be used of the corresponding cache of the target branch BUT the results shall not be written into the cache. Rationale: in case of dependabot only a few new jars will be introduced, the penalty of downloading these are minimal. This way the cache will remain for a significant percentage stable, hopefully 98-99 %

Finally -to avoid race conditions- de master/develop/develop-xxxx caches will only be filled by the 'linux' build as this one has been most stable over time. There are no/very little platform dependent jars anyway.

so the cache key will not depend on
- platform/architecure
- hash over pom files
- individual branch names.
- moon phase

@line-o @reinhapa 

(provided that it can work @duncdrum )

**Comment by duncdrum** [comment]
Yes I have a rough plan, just reading up on some details. But that's the main gist. 

The race conditions require a different fix, but that's more straight forward.
